keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1197 @@
+the control plane and the DRA drivers) in the future, but currently this is out
+of scope.
+
+Embedded inside the NodeResourceSlices is the list of resources managed by the
+driver. Typically there is a single object per driver, but this could also get
+spread out across multiple if a single object would be too large.
+
+Here is an example where the driver provides one discrete card which has
+a certain amount of compute units and RAM that can be reserved:
+
+```
+kind: NodeResourceSlice
+apiVersion: resource.k8s.io/v1alpha2
+...
+nodeName: worker-1
+driverName: cards.dra.example.com
+resources:
+- name: CardA
+  attributes:
+  - name: meta
+    metadata:
+      versions:
+      - name: API
+        version: "v3.2.1" # Must comply with semver v2.0.0.
+      labels:
+      - name: SupportsPowerSaving
+      - name: Color
+        value: "Red" # arbitrary string
+      values:
+      - name: LengthInCentimeters
+        quantity: 100 # int64
+  - name: RAM
+    counter:
+      count: 1000000 # int64, could also be a resource.Quantity for readability
+  - name: ComputeUnits
+    counter:
+      count: 100
+```
+
+The `metadata` and `counter` fields are mutually exclusive. Their content is
+the *attribute type* of a *semantic model*. That model defines the structure of
+the information and implements the logic that compares requests (defined below)
+against these values.
+
+More such fields and models may get added in the future. When old clients
+serialize into their known version of a NodeResourceSlice, they'll encounter
+an `attributes` entry with no known field set. This tells them that they
+cannot handle the object because it was extended.
+
+The `metadata` field will be used by CEL expressions. Because it contains one
+list per type of value, the CEL expression doesn't need to convert from string
+to those types during evaluation, which makes it fully type-safe. Validation
+already happens when publishing the `NodeResourceSlice`.
+
+`metadata` is an example of an attribute which is qualitative, not
+quantitative, and one which contains more than one value. Once we are sure that
+this is how we want to handle meta data and that there is no confusion with
+other qualitative attributes, then the struct can also be lifted up into a
+`metadata` field in the `resources` entries.
+
+In the following example, a different driver instead describes a pool of
+identical cards:
+
+```
+kind: NodeResourceSlice
+apiVersion: resource.k8s.io/v1alpha2
+...
+nodeName: worker-1
+driverName: pool.dra.example.com
+resources:
+- name: PoolOfLargeCards
+  attributes:
+  - name: meta
+    metadata:
+      labels:
+      - name: Configuration
+        value: Large
+  - name: Number
+    counter:
+      count: 8
+```
+
+If a driver needs to reduce resource capacity, then there is a risk that a claim gets
+allocated using that capacity while the kubelet is updating the
+NodeResourceSlices. The implementations of semantic models must handle
```

Comment:
The scheduler could also send multiple pods to a Node before observing the update of NodeResourceSlices.
Do you have any plans to avoid that?

Reply:
Patrick will be updating this to reflect this (if it's not already below), but in DRA there is already the concept of an AllocationResult. This will be captured along with the scheduler decisions. Consider it "allocation intent". The scheduler can construct the "current capacity" by looking at the "base capacity" and accumulating all the allocations. Those are persisted to survive scheduler restarts.

Coming soon in this KEP is the addition of a periodic reporting of "allocation status" from the source of truth (typically on-node driver). If our "allocation intent" and the "allocation status" do not match we can flag that there is an issue (subject to lag of course).